Rewrite mirror configuration documentation

flyinghyrax opened the pull request:
This is a rewrite of the documentation page `docs/mirror_configuration.md` in preparation for fixing the default value issue in #990. I set out to make sure all the options and defaults were documented and got a bit carried away. If this is too drastic of a change, I'm happy to backpedal to an update that _only_ adds missing options and defaults.

### Summary of Major Changes

- Add reference for valid `[mirror]` options that were not listed on the mirror configuration page, e.g.:
    - `storage-backend`
    - `digest_name`
    - `cleanup`
    - `download-mirror-no-fallback`
    - `keep_index_versions`
- Document the expected default value for every option.
    - This will temporarily continue to disagree with reality for `master`, `global-timeout`, `timeout`, `workers`, and `hash-index` when a user gives a config file that is missing any of those options.
- Rewrite each options short description:
    - Move each option's "type" and default value out of the short description and into a field list, so these are displayed consistently across all options.
    - Trim short descriptions to be as concise as possible, moving additional details into separate paragraphs below the type/default.
    - Remove example code blocks that only show a single option by itself.
    - Make extensive (or perhaps excessive?) use of cross-references.
- Create new example snippets that show related options being used together.
- Add documentation for how different options affect the mirror directory structure
    - I personally would've found this useful when I first started using Bandersnatch, since it was hard for me to see what options controlled what directories and files in the output. But I'm also happy to move these elsewhere if you think this is too much detail for this page.
- Embed the content of `src/bandersnatch/default.conf` for user reference
    - Since it's pulled directly from source it should always be up to date as long as the docs get re-built.

### Page Organization

1. Introduction
2. Examples
3. Option Reference
4. Default Config File

The reference for each option is organized like this:

1. Option name (header)
2. Short description (preferably single sentence)
3. Option type and default value
4. Detailed description, notes, and references

I am hoping this makes the page more skimmable, but am unsure.

flyinghyrax commented:
`mdformat` is breaking on MyST syntax. I tried changing the syntax plugin in pre-commit configuration from `mdformat-gfm` to `mdformat-myst` which fixed some problems, but continues to break admonition blocks per this issue: <https://github.com/executablebooks/mdformat-myst/issues/13>. `mdformat-myst` doesn't support MyST's `colon_fence` extension, despite "myst-parser", "mdformat", and "mdformat-myst" all being maintained by the same organization, which seems... kinda weird.

I'm going to try using [code fence syntax for directives](https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html#directives-a-block-level-extension-point) and see if that's enough to smooth things over. (Writing an mdformat plugin feels excessive.)

cooperlees commented on docs/mirror_configuration.md:
````diff
@@ -1,272 +1,408 @@
 
-Example:
+Enable cleanup of legacy simple directories with non-normalized names.
 
-```ini
-[mirror]
-proxy=http://myproxy.com
+:Type: boolean
+:Default: false
+
+Bandersnatch versions prior to 4.0 used directories with non-normalized package names for compatability with older versions of pip. Enabling this option checks for and removes these directories.
+
+```{seealso}
+[Python Packaging User Guide - Names and Normalization](https://packaging.python.org/en/latest/specifications/name-normalization/)
````
I should prob remove this one day.

flyinghyrax replied:
Would it make sense to go ahead and indicate that it's deprecated in the docs? (Though I suppose technically that requires a version bump even if its just the documented status that changes?)

cooperlees replied:
Yeah, will not hurt.

cooperlees commented on docs/mirror_configuration.md:
````diff
@@ -1,272 +1,408 @@
-```
+Recommendations:
+
+- leave the default of 3 to avoid overloading the pypi master
````
```suggestion
- leave the default of 3 to avoid overloading the pypi master and maybe even your own internet connection
  - if you see timeouts and have a slow connection, try lowering this setting
```

flyinghyrax replied:
Noted, haha

cooperlees commented on docs/storage_options.md:
```diff
@@ -6,6 +6,8 @@ Bandersnatch was originally developed for POSIX file system. Bandersnatch now su
 - [Amazon S3](https://aws.amazon.com/s3/)
 - [OpenStack Swift](https://docs.openstack.org/swift/latest/)
 
+(storage-backend-filesystem)=
+
```
Is this work in progress?

flyinghyrax replied:
Aha, this is an explicit MyST ["explicit cross-reference target"](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#creating-explicit-targets) - `mdformat` then added a line between the target and the heading.

I added this when figuring out referencing across documents. I think I can remove it and use the auto-generated heading anchors instead. (MyST has what feels like dozens of ways to make cross references...)

cooperlees reviewed (commented):
Thanks. Looks good to me so far. Just left some minor comments.

flyinghyrax commented on docs/mirror_configuration.md:
```diff
@@ -1,272 +1,408 @@
+directory = /srv/pypi-filtered
+simple-format = ALL
+release-files = false
+root_uri = https://files.pythonhosted.org/
```
I think the current behavior is that if `release-files` is false, `root_uri` will default to `https://files.pythonhosted.org/`, so not required per-se.

I'll have to check, but I suspect setting `root_uri = ` and `release-files = false` would still generate the index files with relative URLs, so I guess it's a matter of what we think the most intuitive default behavior is for `release-files`?

cooperlees replied:
Yup, lets just document what happens in this PR and if we feel we can discuss changing it in a dedicated issue.

flyinghyrax commented on docs/mirror_configuration.md:
```diff
@@ -1,272 +1,408 @@
 
-The mirror json setting is a boolean (true/false) setting that indicates that
-the json packaging metadata should be mirrored in addition to the packages.
+; per-coroutine time limit
```
Based on the existing docs and `default.conf` I believe it's the maximum runtime for each aiohttp coroutine, but this would be a good one to track down in the source to confirm that and perhaps explain more clearly.

cooperlees replied:
Only if you have time, won't block the PR on it.

flyinghyrax commented on docs/mirror_configuration.md:
```diff
@@ -1,272 +1,408 @@
+### `simple-format`
 
-The timeout value is an integer that indicates the maximum number of seconds for web requests.
+The formats to generate for project index files.
```
Good idea!

flyinghyrax commented on docs/mirror_configuration.md:
```diff
@@ -1,272 +1,408 @@
+### `proxy`
 
-The root_uri is a string containing a uri which is the root added to relative links.
+Use an HTTP proxy server.
```
Good to know, will add!

flyinghyrax commented on docs/mirror_configuration.md:
````diff
@@ -1,272 +1,408 @@
+It can also be used with 7zip to create split archives for transfers:
+
+```console
+7za a -i@"/srv/pypi/mirrored-files" -spf -v100m path_to_new_zip.7z
````
Same - I copied this from the existing page content. 7-zip apparently has all sorts of interesting dials and knobs - I guess I've missed a lot just using the GUI version to open tar files. 😆

flyinghyrax commented:
1. I've updated this to match the current config implementation rather than the aspirational one. All options are marked as either required or optional, and I listed the currently required options at the top of the document.
2. I added info about SOCKS support to the `proxy` option, but based on my reading of the source SOCKS is only supported when using environment variables for the proxy URL. It might be worth tweaking the code that checks for SOCKS URLs to check the proxy option in the config as well as the environment variables.
3. I originally thought the output files created by the `json` option were basically unused, but it appears that the `verify` subcommand relies on the files being in `<directory>/web/json`, making enabling `json` a requirement for running `verify`. I will try to validate this experimentally.

I would still like to track down how `global-timeout` is used so I can describe that more accurately. I also plan to run each of the example configs to make sure they work, and verify that I've identified all the 'required' options.

cooperlees commented:
1. Nice - Will review once you mark it as ready + CI passes (just approved it to run)
2. Opened an issue - Thanks for hi-lighting that.
3. The verify option should download the JSON if missing, but probably worth noting it's needed for faster verify.

I'm happy to merge as is and you can do smaller PRs addressing the global-timeout etc. etc. if you like - but happy to do it how ever you want :)
- Just want to make it easier for you. This is already a large improvement as is. Thanks!

flyinghyrax commented:
I tested the example configurations and did find some corrections, so I'm glad I did that. 😄 I also corrected the default value for `diff-file` and added an entry to `CHANGES.md`. I feel it is ready for review.

During the config testing I stumbled into a few things:

* The `mirror` subcommand uses a default value of `filesystem` for `mirror.storage-backend`, but `verify` doesn't provide a default and crashes if `storage-backend` isn't explicitly set. That is something I can easily fix along with the rest of the configuration refactoring.
* I couldn't get `verify` to work without `web/json/` already being populated (i.e. `mirror.json` enabled). It crashes if the 'json' folder doesn't exist, and no-ops if the folder exists but is empty. I haven't yet checked if there are any open issues for this.
* `mirror.master` and `mirror.download-mirror` both seem sensitive to trailing forward slashes (e.g. `https://pypi.org` vs `https://pypi.org/`. I don't know if that's a problem, but it was at least unexpected.
* And one last weird one: running `bandersnatch --debug -c ... mirror` consistently failed if `mirror.directory` wasn't already populated. I didn't try to debug it (heh) and figured I would get together a minimal reproducible example or unit test to replicate it before opening an issue.

cooperlees reviewed (approved):
Many thanks for this PR and all the findings. Feel free to open Issues or just submit PRs if time permits!
